### DISCUSSION TITLE
Are extensions expected to work?

### DISCUSSION THREAD

**Discussion by dhirschfeld** [jupyterlab/retrolab#170] [Q&A] [answered]
I'm trying to display a [`perspective`](https://perspective.finos.org/docs/md/python.html#perspectivewidget) widget in an `ipywidgets.Output`. It works in JupyterLab but seems to get stuck in `retrolab`.

![image](https://user-images.githubusercontent.com/881019/123724880-7143d480-d8d0-11eb-9b73-31bfd456b380.png)

...and I get a bunch of js errors:

![image](https://user-images.githubusercontent.com/881019/123725012-ac460800-d8d0-11eb-953a-d16d668d9bc8.png)

**Comment by jtpio** [comment, marked as answer]
Thanks @dhirschfeld for the report.

Widgets and existing JupyterLab extensions are supported. So if `perspective` behaves like a normal widget then it should work.

The console errors seem to be coming from other extensions that expect to find tokens only exposed in JupyterLab. Usually it's a better practice to not `require` these tokens in extensions, but instead make them `optional` so they don't fail to activate in alternative lab distribution like RetroLab (retro doesn't have a status bar or layout restorer).

**Reply by dhirschfeld** [reply]
> So if perspective behaves like a normal widget then it should work.

I wouldn't be surprised if it didn't. On the token issue, that sounds like a problem with the extensions so I'll close this as asked and answered - thanks!

**Reply by jtpio** [reply]
> with the extensions so I'll close this as aksed and answered

Yes exactly. 

Maybe we could actually improve the lab documentation to tell extension authors to try to make their plugins use `optional` dependencies as much as possible.

**Reply by dhirschfeld** [reply]
That would probably be well worthwhile. Maybe the cookiecutter needs updating too in case it's not doing the right thing?
https://github.com/jupyterlab/extension-cookiecutter-ts/

**Reply by jtpio** [reply]
The cookiecutter already sets the `ISettingRegistry` as optional, and doesn't have other dependencies at the moment so it should be good.

**Reply by timkpaine** [reply]
<moved to separate answer>

**Comment by timkpaine** [comment]
1. @dhirschfeld  there are no perspective errors in the provided screenshot, if you have some in the console I can take a look
2. as [with other partial jupyterlab frontend implementations](https://github.com/voila-dashboards/voila/pull/481), there are probably some dependencies which constitute a "normal" jupyterlab install that are missing. @jtpio The statements "Widgets and existing JupyterLab extensions are supported." and "The console errors seem to be coming from other extensions that expect to find tokens only exposed in JupyterLab" directly contradict one another. Also contradicts the description itself: "Extensions built for Jupyterlab should work as long as the components they depend on are part of this classic interface (for install RetroLab doesn't have the left, right and bottom areas)."

**Reply by timkpaine** [reply]
Based on the list I did at the time for `voilà`, I think `retrolab` fails to expose a "normal" jupyterlab dependency surface and therefore I think the answer to your question is "no, extensions are not necessarily compatible". It looks like its missing:
- `@jupyterlab/outputarea`
- `@lumino/signaling`
- `@lumino/virtualdom`
- `@lumino/algorithm`
- `@lumino/domutils`

From past experience with `voilà`, any "normal" jupyterlab extension that interacts with `@lumino` in a deeper way, or `@jupyterlab` in particular for `rendermime` extensions will have trouble. `perspective` is a "normal" jupyterlab extension but uses lumino directly and exports a rendermime for `json`, `csv`, and `arrow` files.

**Reply by timkpaine** [reply]
prior analysis which is suspect is relevant:
https://github.com/voila-dashboards/voila/pull/481#issuecomment-585412635

**Reply by dhirschfeld** [reply]
> From past experience with voilà, any "normal" jupyterlab extension that interacts with @lumino in a deeper way, or @jupyterlab in particular for rendermime extensions will have trouble

Yeah, my comment about not being surprised if it didn't work was just in regards to the fact that's it's not a "simple" widget and is therefore more likely to push the boundaries of what is required.

I actually tried to get `perspective` to work in `voila` first without any luck (so far). Since your PR is merged do you expect `perspective` should work with `voila`?

**Reply by timkpaine** [reply]
https://github.com/finos/perspective/issues/1454

**Reply by dhirschfeld** [reply]
Thanks @timkpaine - I'm now subscribed to the issue & pr!

I'm just starting to try out `perspective` and it seems great. I do think a killer feature would be `voila` support as it makes distributing the results of your notebooks easier than having to tell people to just click the execute & run all button.

That's kind-of why I'm looking into `retrolab` too - if a user (non- developer/analyst) does have to run a notebook, `retrolab` might be a less intimidating entrypoint than the full-blown JupyterLab UI.

**Reply by timkpaine** [reply]
it used to support voila, im sure it will again given our extensive use of the perspective+voila combination for business users at JPMC (for exactly the scenarios you describe, no- or low-code users like sales people who want to see the results of analysts/quants/business intelligence/trader-developers). I would think the barrier to supporting arbitrary extensions is even lower for retrolab since it includes far more of the jupyterlab assets than voila, so we'll just need to track down which parts are missing and add them

**Comment by timkpaine** [comment]
An alternative answer which maybe @jtpio can clarify is from the README:
"Existing federated JupyterLab extensions listed via..."

So "No, only some federated extensions are supported".
